src/python/pants/backend/python/dependency_inference/module_mapper.py

```diff
@@ -0,0 +1,101 @@
+from pants.engine.rules import rule
+from pants.engine.selectors import Get, MultiGet
+from pants.engine.target import Targets
+from pants.source.source_root import AllSourceRoots
+
+
+@dataclass(frozen=True)
+class PythonModule:
+    module: str
+
+    @classmethod
+    def create_from_stripped_path(cls, path: PurePath) -> "PythonModule":
+        module_name_with_slashes = (
+            path.parent if path.name == "__init__.py" else path.with_suffix("")
+        )
+        return cls(module_name_with_slashes.as_posix().replace("/", "."))
+
+    def address_spec(self, *, source_root: str) -> AscendantAddresses:
+        """The spec for all candidate targets which could feasibly own the module.
+
+        This uses AscendantAddresses because targets can own files in subdirs (e.g. rglobs). We also
+        use the package path, e.g. `helloworld/util/__init__.py`, rather than the module path to
+        ensure that we capture all possible targets. It is okay if this directory does not actually
+        exist.
+        """
+        module_name_with_slashes = self.module.replace(".", "/")
+        return AscendantAddresses(
+            directory=PurePath(source_root, module_name_with_slashes).as_posix()
+        )
+
+    def possible_stripped_paths(self) -> Tuple[PurePath, ...]:
+        """Given a module like `helloworld.util`, convert it back to its possible paths.
+
+        Each module has either 2 or 4 possible paths. For example, given the module
+        `helloworld.util`:
+
+        - helloworld/util.py
+        - helloworld/util/__init__.py
+        - helloworld.py
+        - helloworld/__init__.py
+
+        The last two possible paths look at the parent module, if any. This is to accommodate `from`
+        imports, where we don't care about the specific symbol, but only the module. For example,
+        with `from typing import List`, we only care about `typing`.
+        """
+        module_name_with_slashes = PurePath(self.module.replace(".", "/"))
+        possible_paths = [
+            module_name_with_slashes.with_suffix(".py"),
+            module_name_with_slashes / "__init__.py",
+        ]
+        parent = module_name_with_slashes.parent
+        if str(parent) != ".":
+            possible_paths.extend([parent.with_suffix(".py"), parent / "__init__.py"])
+        return tuple(possible_paths)
+
+
+class PythonModuleOwners(DeduplicatedCollection[Address]):
+    """The targets that own a Python module."""
+
+    sort_input = True
+
+
+@rule
+async def map_python_module_to_targets(
+    module: PythonModule, source_roots: AllSourceRoots
+) -> PythonModuleOwners:
+    unfiltered_candidate_targets = await Get[Targets](
+        AddressSpecs(module.address_spec(source_root=src_root.path) for src_root in source_roots)
+    )
+    candidate_targets = tuple(
+        tgt for tgt in unfiltered_candidate_targets if tgt.has_field(PythonSources)
+    )
+    sources_per_target = await MultiGet(
```

Comment:
Could likely do some pre-filtering here with `any_matches_filespec` to eliminate hydration of targets that couldn't own the file. Not a blocker though.

Reply:
As discussed over DM, will do in a followup with a benchmark.